(Expo QuickStart) (#1697)
mongodb/docs-realm: source/sdk/react-native/quick-start-expo.txt

Overview
--------

The :github:`Realm Expo template <expo/examples/tree/master/with-realm>`
provides a fully working React Native application that you can use to bootstrap
your app development project with Realm. This documentation covers how to
initialize, build and run the template on both Android and iOS, and the template's

> file structure and core features. 

Reviewer: would be good to add a sentence about what expo is, also linking to the expo docs.

Reviewer: nice addition. see new comment abt how it's rendering

Prerequisites
-------------
- `Node.js version 12 LTS or greater <https://nodejs.org/en/about/releases/>`_

Setup
-----

If you have never used Expo or React Native, follow the `Setting up the
development environment <https://reactnative.dev/docs/environment-setup>`_ guide

> before continuing.

Reviewer: should the user follow the "Expo CLI Quickstart" specifically? would be good to clarify b/c the tabs on the linked page have 2 flows, which are quite distinct

Next, before you initialize the template, you must install the `expo cli
<https://docs.expo.dev/workflow/expo-cli/>`_ by running the following command:

.. code-block:: shell
  
   npm install -g expo-cli

Initialize the Template
-----------------------

> Use ``expo init``  with the flag ``--template @realm/expo-template-ts`` to initialize a
> React Native application using the Realm Expo template. 
> 
> Run the following command in your terminal: 

Reviewer: ```suggestion
Use the Expo CLI to initialize a
React Native application using the Realm Expo template. Run the following command in your terminal: 
```

Author: Decided to kind of use your wording, but kept some of my wording too.

.. code-block:: shell
  

>    expo init MyAwesomeRealmApp --template @realm/expo-template-ts

Reviewer: any reason why it's just including the TS template here? 

i see there's also a JS version - https://www.npmjs.com/package/@realm/expo-template-js

Author: Added a JS version now 👍

Reviewer: looks like there's an issue w formatting for both tabs:
![Screen Shot 2022-03-30 at 11 45 42 AM](https://user-images.githubusercontent.com/90647379/160876328-0e5aa870-bdb7-406e-ae9f-7051757b857f.png)

Explore The File Structure
--------------------------

The relevant files are as follows:

.. cssclass:: config-table
.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - File
     - Purpose

   * - Task.ts
     - A typescript file that defines a Task :ref:`object schema
       <react-native-object-schemas>` and returns a ``RealmProvider`` and the

>        ``Realm React Hooks``.

Reviewer: why is this highlighted?

Author: I forget why I originally monospaced it. I've replaced this with just the word 'hooks' now since that's what it returns. Also, included a link to the Realm React documentation at the top of this page now.

      
   * - App.ts
     - The entry point to the application, which contains methods to create,
       update, and delete tasks. The ``RealmProvider`` component wraps around
       the ``App`` component, providing your entire application with access to
       the {+realm+}.

   * - TaskList.tsx

>      - A React component that gets a list of Task objects from the {+realm+} and renders each object as a TaskItem component.

Reviewer: ```suggestion
     - A React component that gets a list of Task objects from the {+realm+} and renders each object as a ``TaskItem`` component.
```

Author: Fixed locally 👍

   * - TaskItem.tsx
     - A React component that displays the task's description, a checkbox to
       toggle the task's completion status, and a button to delete the task.

Build and Run the Application
-----------------------------

To run your application on an :android:`Android emulator <studio/run/emulator>`,
run the following command from your application root:

.. code-block:: shell
   
   npm run android

To run your application on an :apple:`iOS simulator
<documentation/xcode/running-your-app-in-the-simulator-or-on-a-device>`, run the
following command from your application root:

.. code-block:: shell
   
   npm run ios

When the application builds, you will see something similar to the following image:

.. image:: /images/realm-expo-template.png
   :width: 400px
   :alt: Realm Expo Template on an iOS Simulator

When you run the commands above, the Expo `dev-client
<https://docs.expo.dev/development/getting-started/>`_ will start automatically.
If you close your Expo ``dev-client`` and want to restart it without waiting for
the native build process again, you can simply run: 

.. code-block:: shell
   
   npm start 

However, if you made any `changes to the native code

> <https://docs.expo.dev/workflow/customizing/>`_, you will have to rerun ``npm
> run android`` or ``npm run ios`` to rebuild your native iOS/Android app.

Reviewer: just a thought to streamline a bit. i think any dev would understand, but might be unclear.

```suggestion
<https://docs.expo.dev/workflow/customizing/>`_, you will have to rerun ``npm
run android | ios`` to rebuild your native iOS/Android app.
```

Author: Resolving, as I think this is more a personal preference, and I prefer the current form.

Deploying Your Application
--------------------------

> To deploy your application, you can use the `Expo Application Services (EAS)

Reviewer: i believe there are other ways to deploy w expo, even if this is the easiest. 

would be good to rephrase to make clear this is just one option. 

and then additionally link to expo deployment docs in another sentence https://docs.expo.dev/distribution/introduction/

Author: Fixed w/ your suggestions 👍

<https://docs.expo.dev/eas/>`_, a suite of deeply integrated cloud services for

> Expo and React Native apps. EAS allows you to compile and :wikipedia:`sign

Reviewer 2: Theres another cool feature about EAS that we could feature.  A developer can ask the EAS to build the app dev client, which can be downloaded on their mobile device.  With the downloaded dev client, they can scan the QR code shown when they start the dev-server on their development machine, which will then allow them to start making changes to their app code which will be automatically shown on their device.  Similar to Expo Go.  There is a bit of setup involved, especially with iOS (dealing with registering UDID of development devices).  This has the advantage that developers don't need to configure xcode or android studio

Author: That's an interesting feature, but it feels unrelated to the subsection 'Deploying Your Application'. That seems like more of an alternative to the commands that are currently in 'Build and Run the Application'. I'm happy to include it there, or disclude it altogether from this guide, but I don't think it would belong in this section. What do you think @takameyer ?

Author: Is this what you're referring to https://docs.expo.dev/guides/configuring-updates/ when you mean 'update an app without making an app store release'. Going to merge this as-is for now to get this out to users, but can you create a list of links you would want Realm Expo users to use, and then we can find a place in the docs for it? (Maybe just at the end of this page as a see-also or maybe we can create a separate page of next steps that a user might take)

Overview
--------

The :github:`Realm Expo template <expo/examples/tree/master/with-realm>`
provides a fully working React Native application that you can use to bootstrap
your app development project with Realm. This documentation covers how to
initialize and work with the Realm Expo template.

The Realm Expo template uses:

-  `Expo <https://docs.expo.dev/>`_, a framework to develop, build and deploy
   React Native applications quickly.
- :npm:`Realm React <package/@realm/react>`, an NPM package that provides an
   easy-to-use API to perform common Realm operations, such as querying or
   writing to a {+realm+} and listening to realm objects. Read the :ref:`Realm
   React documentation <react-native-use-realm-react>` to learn more about

>    working with it.

Reviewer: this bullet point is rendering as a definition list. looks like there might be an extra space on all the lines after the first?
![Screen Shot 2022-03-30 at 11 51 24 AM](https://user-images.githubusercontent.com/90647379/160877495-0694211b-6bc6-4976-9182-0852b2e87c0f.png)
